internal/tofu/provider_for_test_framework.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package tofu
+
+import (
+	"github.com/opentofu/opentofu/internal/addrs"
+	"github.com/opentofu/opentofu/internal/configs/hcl2shim"
+	"github.com/opentofu/opentofu/internal/providers"
+	"github.com/zclconf/go-cty/cty"
+)
+
+var _ providers.Interface = providerForTest{}
+
+// providerForTest is a wrapper around real provider to allow certain resources to be overridden
+// for testing framework. Currently, it's used in NodeAbstractResourceInstance only in a format
+// of one time use. It handles overrideValues and plannedChange for a single resource instance
+// (i.e. by certain address).
+// TODO: providerForTest should be extended to handle mock providers implementation with per-type
+// mocking. It will allow providerForTest to be used for both overrides and full mocking.
+// In such scenario, overrideValues should be extended to handle per-type values and plannedChange
+// should contain per PlanResourceChangeRequest cache to produce the same plan result
+// for the same PlanResourceChangeRequest.
+type providerForTest struct {
+	// It's not embedded to make it safer to extend providers.Interface
+	// without silently breaking providerForTest functionality.
+	internal providers.Interface
+	schema   providers.ProviderSchema
+
+	overrideValues map[string]cty.Value
+	plannedChange  *cty.Value
+}
+
+func (p providerForTest) ReadResource(r providers.ReadResourceRequest) providers.ReadResourceResponse {
+	var resp providers.ReadResourceResponse
+
+	resSchema, _ := p.schema.SchemaForResourceType(addrs.ManagedResourceMode, r.TypeName)
+
+	resp.NewState, resp.Diagnostics = hcl2shim.ComposeMockValueBySchema(resSchema, r.ProviderMeta, p.overrideValues)
+	return resp
+}
+
+func (p providerForTest) PlanResourceChange(r providers.PlanResourceChangeRequest) providers.PlanResourceChangeResponse {
+	if r.Config.IsNull() {
+		return providers.PlanResourceChangeResponse{
+			PlannedState: r.ProposedNewState, // null
+		}
+	}
+
+	if p.plannedChange != nil {
```

Comment:
I don't really understand this part of the `planned change` and how it affects the overridden resource. Can you explain a little?

Reply:
Sure!

In general, `providerForTest` uses configuration values (from `values` / `outputs` field in `override_` block) and automatically generates some missing computed values. When doing so, it uses some rules on how the values should be generated and the value is random for strings.

After this, tofu checks if the provider generated the correct response. It's the point when the logic breaks without this `if`. We need to use the same generated value for multiple sequential provider calls (i.e. plan is called twice).

Currently, `providerForTest` is meant to be used once inside the `NodeAbstractResourceInstance` so we just pass the previously planned value to the `providerForTest`.

It will be changed to the proper cache when extending the `providerForTest` to support `mock_provider` feature. This one: https://github.com/opentofu/opentofu/issues/1155

I left a comment explaining it a bit more here: https://github.com/opentofu/opentofu/blob/override-for-testing/internal/tofu/provider_for_test_framework.go#L21

Reply:
Thanks for the explanation, this is much clearer! 🙏